### PR TITLE
Add caching to WP_Theme_JSON_Resolver_Gutenberg::get_user_data_from_custom_post_type()

### LINKED ISSUE #36574: WP_Theme_JSON_Resolver_Gutenberg::get_user_data_from_custom_post_type uses uncached SQL queries
### Description

We noticed a few performance related issues in big multisite WP site related to the uncached queries used in `WP_Theme_JSON_Resolver_Gutenberg::get_user_data_from_custom_post_type()`. While there's some caching done at the class level, it'd be good to use object caching to avoid unnecessary SQL queries triggered by the call to `wp_get_recent_posts`.

### Step-by-step reproduction instructions

Please run the following snippet in a WP install with Gutenberg installed.

### Screenshots, screen recording, code snippet

```
<?php
require __DIR__ . '/wp-load.php';

$__queries = array();
$__func = function( $query ) use ( &$__queries ) {
        if ( preg_match( '#post_type = \'wp_global_styles\'#', $query ) ) {
                $__queries[] = $query;
        }
        return $query;
};
add_filter( 'query', $__func );
for ( $i = 0; $i < 10; $i++ ) {
        WP_Theme_JSON_Resolver_Gutenberg::get_user_data();
        WP_Theme_JSON_Resolver_Gutenberg::clean_cached_data();
}
var_dump($__queries);
```
Ideally, we should only do the SQL query once.

### Environment info

- WordPress 5.8.2
- Gutenberg 11.8.3 (the latest versions are also affected, and also add more similar code in `gutenberg_add_active_global_styles_link`)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by xknown**
Refactor
`WP_Theme_JSON_Resolver_Gutenberg::get_user_data_from_custom_post_type`
to add caching to the `wp_global_styles` post type lookup, and also to
remove the duplicate code added in
`gutenberg_add_active_global_styles_link`.

Fixes #36574

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 1]
```diff
@@ -170,33 +170,47 @@ public static function get_theme_data() {
...
+	 *                                    If empty, it defaults to the current theme.
+	 * @param bool     $should_create_cpt Whether a new CPT should be created if no one was found.
+	 *                                    False by default.
+	 * @param array    $post_status_filter Filter CPT by post status.
+	 *                                    ['publish'] by default, so it only fetches published posts.
 	 *
 	 * @return array Custom Post Type for the user's origin config.
 	 */
-	private static function get_user_data_from_custom_post_type( $should_create_cpt = false, $post_status_filter = array( 'publish' ) ) {
+	public static function get_user_data_from_custom_post_type( $theme, $should_create_cpt = false, $post_status_filter = array( 'publish' ) ) {
```
Comment (Mamaduka) [inline lib/class-wp-theme-json-resolver-gutenberg.php:207]: Hi, @xknown

What's the reason for adding the `$theme` argument?

I understand visibility change is required for unit tests, but can we use `::get_user_data()` like your initial test in the issue?

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 2]
```diff
@@ -170,33 +170,47 @@ public static function get_theme_data() {
...
+				array(
+					'taxonomy' => 'wp_theme',
+					'field'    => 'name',
+					'terms'    => $theme->get_stylesheet(),
 				),
-			)
+			),
 		);
 
+		$cache_key = sprintf( 'wp_global_styles_post_type_%s', md5( serialize( $args ) ) );
```
Comment (Mamaduka) [inline lib/class-wp-theme-json-resolver-gutenberg.php]: I think we can drop the `post_type` part from the cache key.

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 1 above)
Comment (xknown) [reply lib/class-wp-theme-json-resolver-gutenberg.php:207]: The visibility/signature change is required to remove the duplicate code block from `gutenberg_add_active_global_styles_link`.

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 3]
```diff
@@ -170,33 +170,47 @@ public static function get_theme_data() {
...
+		$post_id   = wp_cache_get( $cache_key );
+		if ( false !== $post_id ) {
+			if ( (int) $post_id > 0 ) {
+				return get_post( $post_id, ARRAY_A );
+			} elseif ( -1 === $post_id && ! $should_create_cpt ) {
```
Comment (Mamaduka) [inline lib/class-wp-theme-json-resolver-gutenberg.php]: What do you think about changing this into multiple if statements?

```
if ( (int) $post_id > 0 ) {
     return get_post( $post_id, ARRAY_A );
}

// Special case: '-1' is a results not found.
if ( -1 === $post_id && ! $should_create_cpt ) {
     return $user_cpt;
}
```

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 1 above)
Comment (Mamaduka) [reply lib/class-wp-theme-json-resolver-gutenberg.php:207]: Sorry, I just noticed that part.

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 2 above)
Comment (xknown) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: Removed in 934143c09d

File: lib/class-wp-theme-json-resolver-gutenberg.php (hunk 3 above)
Comment (xknown) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: Changed it in 934143c09d

**Review by Mamaduka — APPROVED** [review]
Thanks for working on this, @xknown.

The code looks good to me. Let's rebase on the current trunk and merge this once all checks are green.

**Comment by xknown** [comment]
> Thanks for working on this, @xknown.
> 
> The code looks good to me. Let's rebase on the current trunk and merge this once all checks are green.

Thanks for the review/feedback. I rebased the PR branch.

**Comment by Mamaduka** [comment]
Thank you, @xknown.

**Comment by youknowriad** [comment]
Is this something that need to be backported in core for 5.9? cc @oandregal @jorgefilipecosta

**Comment by noahtallen** [comment]
A heads up that I cherry-picked this into the gutenberg 12.0.1 point release.

**Comment by noisysocks** [comment]
If we don't backport it then it will be deleted when Gutenberg requires WP 5.9 at minimum, no?

**Comment by youknowriad** [comment]
The changes in ` lib/compat/wordpress-5.9/` yes. We should be more strict with changes made to `lib/compat` if we do want to make changes there without back porting, we need to move the changes to `lib/compat/wordpress-6.0`

**Comment by oandregal** [comment]
@youknowriad Without having looked deeply into the code, I'd think it's best to do it, yes: it changes some public method we're merging in 5.9 for the first time. While it is from a "private/internal" class, I'd rather play it safe.

**Pull request merged**